source/_posts/vue-design-patterns.md

```diff
@@ -0,0 +1,988 @@
+    }
+    .label {
+      color: gray;
+      padding-right: 10px;
+    }
+    </style>
+```
+
+Excuses for my ugly CSS, but that's not the point. Maybe it isn't the ideal Form Builder, but it's useful. The more complex your form type, the more complex your form builder will be. Now you have a Form Builder, which you can use to create forms in each view, but we have one step more, the `FormDirector`. It has the task of structuring our forms in a single file with verbose methods, like `makeLoginForm()`, `makeSignUpForm()`, `makeShopForm()`, and more. Here is our `FormDirector.ts`.
+
+```ts
+    import VInput from "@/components/form/VInput.vue";
+    import type FormBuilder from "./FormBuilder";
+    import z from "zod";
+    
+    export default class FormDirector {
+      builder: FormBuilder;
+      constructor(builder: FormBuilder) {
+        this.builder = builder;
+      }
+      makeLoginForm() {
+        return this.builder
+          .addField({
+            component: VInput,
+            name: "username",
+            type: "text",
+            label: "username",
+            props: {
+              value: "",
+            },
+            validation: z.string().min(10).max(40),
+          })
+          .addField({
+            component: VInput,
+            name: "password",
+            type: "password",
+            label: "password",
+            props: {
+              value: "default password",
+            },
+            validation: z.string().min(10),
+          })
+          .build();
+      }
+    }
+```
+
+To validate our forms, we are using the [zod](https://zod.dev/) library. As we can see, `makeLoginForm()` uses the Form Builder class to create a new app form with a detailed description of each field.
+
+> Code:  [vue design pattern repository](https://github.com/cr0wg4n/vue-design-patterns), execute, and go to **/builder-pattern** route 😎.
+
+
+## Adapter Pattern
+The adapter pattern is a structural design pattern also known as "Wrapper". This pattern allows the interfacing of an existing class used as another interface, like something in the middle of a connection. The perfect real-world example is when you go to another country, you realize that the power plug is different. So, you will need an adapter to get power. That's the magic.
+
+A typical use case is to wrap a 3rd party library, which is used everywhere in your code. Someday, many libraries will be deprecated, or in the worst case, security issues, also known as vulnerabilities. To learn more about it, we have an amazing academy, [https://websecurity-academy.com/](https://websecurity-academy.com/), and a free guide related to [Vue Security Best Practices](https://dev-academy.com/vue-security-best-practices/).
+
+In this case, we are using the [js-cookie](https://github.com/js-cookie/js-cookie) library to interact directly with our Cookies, which is simple and lightweight. The adapter looks like this.
+
+```ts
+    // File: CookiesAdapter.ts
```

Comment:
If you want to add a title to a code snippet here is an example way to do it:

```
``` html A typical DOM-XSS exposure
<h1>Search results for <b>cats</b>:</h1>
```

After triple ticks ``` and the type of the code just write the title (instead of using comments). In this case:

```
```ts CookiesAdapter.ts
```

- [ ] Fix all code snippets titles (remove comments)

Reply:
Done :D